drizzle-orm/type-tests/mysql/set-operators.ts

```diff
@@ -0,0 +1,209 @@
+			.from(users),
+	);
+
+Expect<Equal<{ id: number; homeCity: number }[], typeof exceptTest>>;
+
+const exceptAllTest = await db
+	.select({ id: users.id, homeCity: users.class })
+	.from(users)
+	.exceptAll(
+		db
+			.select({ id: users.id, homeCity: sql<'A' | 'C'>`${users.class}` })
+			.from(users),
+	);
+
+Expect<Equal<{ id: number; homeCity: 'A' | 'C' }[], typeof exceptAllTest>>;
+
+const union2Test = await union(db.select().from(cities), db.select().from(cities), db.select().from(cities));
+
+Expect<Equal<{ id: number; name: string; population: number | null }[], typeof union2Test>>;
+
+const unionAll2Test = await unionAll(
+	db.select({
+		id: cities.id,
+		name: cities.name,
+		population: cities.population,
+	}).from(cities),
+	db.select().from(cities),
+);
+
+Expect<Equal<{ id: number; name: string; population: number | null }[], typeof unionAll2Test>>;
+
+const intersect2Test = await intersect(
+	db.select({
+		id: cities.id,
+		name: cities.name,
+		population: cities.population,
+	}).from(cities),
+	db.select({
+		id: cities.id,
+		name: cities.name,
+		population: cities.population,
+	}).from(cities),
+	db.select({
+		id: cities.id,
+		name: cities.name,
+		population: cities.population,
+	}).from(cities),
+);
+
+Expect<Equal<{ id: number; name: string; population: number | null }[], typeof intersect2Test>>;
+
+const intersectAll2Test = await intersectAll(
+	db.select({
+		id: cities.id,
+	}).from(cities),
+	db.select({
+		id: cities.id,
+	})
+		.from(cities),
+).orderBy(desc(cities.id)).limit(23);
+
+Expect<Equal<{ id: number }[], typeof intersectAll2Test>>;
+
+const except2Test = await except(
+	db.select({
+		userId: newYorkers.userId,
+	})
+		.from(newYorkers),
+	db.select({
+		userId: newYorkers.userId,
+	}).from(newYorkers),
+);
+
+Expect<Equal<{ userId: number }[], typeof except2Test>>;
+
+const exceptAll2Test = await exceptAll(
+	db.select({
+		userId: newYorkers.userId,
+		cityId: newYorkers.cityId,
+	})
+		.from(newYorkers),
+	db.select({
+		userId: newYorkers.userId,
+		cityId: newYorkers.cityId,
+	}).from(newYorkers),
+);
+
+Expect<Equal<{ userId: number; cityId: number | null }[], typeof exceptAll2Test>>;
+
+{
+	const query = db
+		.select()
+		.from(users)
+		.union(
+			db.select()
+				.from(users),
+		)
+		.prepare()
+		.iterator();
+	for await (const row of query) {
+		Expect<Equal<typeof users.$inferSelect, typeof row>>();
+	}
+}
+
+// @ts-expect-error - The select on both sites must be the same shape
+db.select().from(classes).union(db.select({ id: classes.id }).from(classes));
+
+// @ts-expect-error - The select on both sites must be the same shape
+db.select({ id: classes.id }).from(classes).union(db.select().from(classes));
+
+union(
+	db.select({ id: cities.id, name: cities.name }).from(cities),
+	db.select({ id: cities.id, name: cities.name }).from(cities),
+	// @ts-expect-error - The select on rest parameter must be the same shape
+	db.select().from(cities),
+);
+
+union(
```

Comment:
Maybe we should have one more test with columns in a different order. In this case, types for columns in the same position will be different, which should be shown as an error. If you haven't implemented it in the types, it would be ideal to add it as well. If it's something that is not possible or has unnecessary complexity, let's just handle it at runtime

```ts
union(
db.select({ id: cities.id, name: cities.name }).from(cities),
db.select({ name: cities.name, id: cities.id }).from(cities),
)
```
Also let's add this case to every possible union/intersect/except implementation

Reply:
Andrew, unfortunately, implementing this at the type level might not be possible, it would require typescript to have some kind of order on an union type (keyof object). See [this typescript issue](https://github.com/microsoft/TypeScript/issues/13298) for context. I will add a runtime check and throw and error if different key orders are detected.